README.md
# primes

A command-line Clojure program designed to print a multiplication table of prime numbers.

## Usage

`java -jar primes.jar [number of primes]`

## License

Copyright © 2017 David Young-chan Kay

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-03-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-03-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/primes/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/primes/compare/0.1.0...0.1.1


test/primes/core_test.clj
(ns primes.core-test
  (:require [midje.sweet :refer :all]
            [primes.core :refer :all]))

(facts "Prime Counting"
       (future-fact "Prime counting works 100% accurately"
             (num-primes-below 100) => 25)
       (fact "Prime counting works as described by Gauss"
             (num-primes-below 100) => 21))

(facts "Prime Sieve"
       (fact "Can find primes below 30"
             (prime-sieve 30) => [2 3 5 7 11 13 17 19 23 29])

       (fact "Can find primes below 100"
             (prime-sieve 100) => [2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97]))

(facts "Sieve Upper Bound"
       (fact "Can find sieve upper bound, approximately"
             (find-sieve-upper-bound 20) => 100)
       (future-fact "Can find sieve upper bound, exactly"
             (find-sieve-upper-bound 25) => 100))

(facts "Prime Multiplication Table"
       (fact "Can find basic multiplication table"
             (prime-table 10) => {:primes '(2 3 5 7 11 13 17 19 23 29)
                                  :products '((4 6 10 14 22 26 34 38 46 58)
                                              (6 9 15 21 33 39 51 57 69 87)
                                              (10 15 25 35 55 65 85 95 115 145)
                                              (14 21 35 49 77 91 119 133 161 203)
                                              (22 33 55 77 121 143 187 209 253 319)
                                              (26 39 65 91 143 169 221 247 299 377)
                                              (34 51 85 119 187 221 289 323 391 493)
                                              (38 57 95 133 209 247 323 361 437 551)
                                              (46 69 115 161 253 299 391 437 529 667) (58 87 145 203 319 377 493 551 667 841))

                                  }))

(facts "Arg parsing"
       (fact "Can parse an argument"
             (read-int "") => nil
             (read-int nil) => nil
             (read-int "1") => 1))


test/primes/output_test.clj
(ns primes.output-test
  (:require [midje.sweet :refer :all]
            [primes.output :refer :all]
            [clojure.string :as string]))

(def example-table
  {:primes   '(2 3 5 7 11 13 17 19 23 29)
   :products '((4 6 10 14 22 26 34 38 46 58)
               (6 9 15 21 33 39 51 57 69 87)
               (10 15 25 35 55 65 85 95 115 145)
               (14 21 35 49 77 91 119 133 161 203)
               (22 33 55 77 121 143 187 209 253 319)
               (26 39 65 91 143 169 221 247 299 377)
               (34 51 85 119 187 221 289 323 391 493)
               (38 57 95 133 209 247 323 361 437 551)
               (46 69 115 161 253 299 391 437 529 667)
               (58 87 145 203 319 377 493 551 667 841))})

(facts "Pad Entry"
       (fact "Can pad entry"
             (pad-entry 3 "1") => "  1"))

(facts "Format prime table"
       (fact "Can format header"
             (header 3 (:primes example-table)) => "        2   3   5   7  11  13  17  19  23  29"
             )
             
       (fact "Can format row"
             (row 3 "2" (first (:products example-table))) => "  2 |   4   6  10  14  22  26  34  38  46  58")
       
       (fact "Can format prime table"
             (let [expected (string/join "\n"
                                         ["        2   3   5   7  11  13  17  19  23  29"
                                          "    -----------------------------------------"
                                          "  2 |   4   6  10  14  22  26  34  38  46  58"
                                          "  3 |   6   9  15  21  33  39  51  57  69  87"
                                          "  5 |  10  15  25  35  55  65  85  95 115 145"
                                          "  7 |  14  21  35  49  77  91 119 133 161 203"
                                          " 11 |  22  33  55  77 121 143 187 209 253 319"
                                          " 13 |  26  39  65  91 143 169 221 247 299 377"
                                          " 17 |  34  51  85 119 187 221 289 323 391 493"
                                          " 19 |  38  57  95 133 209 247 323 361 437 551"
                                          " 23 |  46  69 115 161 253 299 391 437 529 667"
                                          " 29 |  58  87 145 203 319 377 493 551 667 841"])]
                   (format-prime-table example-table) => expected)))



src/primes/core.clj
(ns primes.core
  [:require [primes.output :refer [format-prime-table]]]
  (:gen-class))

(defn num-primes-below
  "How many primes are below N?

  Note: this implementation is an APPROXIMATION, based on Gauss & Legendre
  The good news is that it under-counts, so we will never end up with too low of a ceiling / too few primes. We will, however, end up doing a bit of extra computation in prime-sieve.
  https://en.wikipedia.org/wiki/Prime-counting_function"
  [n]
  (int (/ n
          (Math/log n))))

(def sieve-bounds (for [i (range 1 (Math/log10 Integer/MAX_VALUE))]
                    (let [sieve-bound (int (Math/pow 10 (Math/ceil i)))
                          n-primes (num-primes-below sieve-bound)]
                      [n-primes sieve-bound])))

(defn square [x]
  (* x x))

(defn find-sieve-upper-bound
  "Given a number of primes to find, K, return the upper bound we will need to sieve for.
  Note: this could definitely be optimized. We use a lookup table that only has every power of 10 present."
  [k]
  (->> sieve-bounds
       (filter (fn [[n-primes sieve-bound]]
                 (> n-primes k)))
       first
       last))

(defn prime-sieve
  "Use a sieve to find all primes up to N.
  Implemented using Sieve of Eratosthenes
  https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
  Very imperative style for now. Can come back later and make it more functional."
  [n]
  ;; Input: an integer n > 1.
  ;;  
  ;;  Let A be an array of Boolean values, indexed by integers 2 to n, initially all set to true.
  ;; Note: into-array likely less performant than make-array. Optimize this if necessary.
  (let [array (into-array (repeat n true))]
    ;;  for i = 2, 3, 4, ..., not exceeding √n:
    (doseq [i (range 2 (Math/sqrt n))]
      ;; if A[i] is true:
      (when (aget array i)
        ;;      for j = i2, i2+i, i2+2i, i2+3i, ..., not exceeding n:
        (doseq [j (range (square i) n i)]
          ;;        A[j] := false.
          (aset array j false))))

    ;;  Output: all i such that A[i] is true.
    (->> array
         (map-indexed (fn [i x] [i x]))
         (drop 2)
         (filter (fn [[i x]] (true? x)))
         (map first))))

(defn get-primes [k]
  "Find k primes"
  (let [upper-bound (find-sieve-upper-bound k)]
    (take k (prime-sieve upper-bound))))

(defn prime-table
  "Multiplication table of the first k prime numbers."
  [k]
  (let [primes (get-primes k)]
    {:primes primes
     :products (for [a primes]
                 (for [b primes]
                   (* a b)))}))
;; Notes 
;; ● Consider complexity. How fast does your code run? How does it scale? 
;; ● Consider cases where we want ​N ​ primes. 
;; ● Do not use the Prime class from stdlib (write your own code). 
;; ● Write tests. Try to demonstrate TDD/BDD. 
;; ● If you’re using external dependencies please specify those dependencies and how to install them. 
;; ● Please ​package your code​, OR include running instructions. 

(defn print-prime-table
  "Prints multiplication table of the first k prime numbers."
  [k]
  (println (format-prime-table (prime-table k))))

(defn read-int [s]
  (try
    (Integer/parseInt s)
    (catch NumberFormatException e
      nil)))
    
(defn -main
  "Prints multiplication table of the first 10 prime numbers."
  [& args]
  (print-prime-table (or (read-int (first args))
                         10)))


src/primes/output.clj
(ns primes.output
  [:require [clojure.string :as string]])

(def vertical-divider "|")
(def horizontal-divider "-")

(defn repeat-string [c n]
  (apply str (repeat n c)))

(defn header-divider [cell-width num-primes]
  (str
   (repeat-string " " (inc cell-width))
   (repeat-string horizontal-divider (inc (* num-primes (inc cell-width))))))

(defn pad-entry [cell-width e]
  (let [len (count (str e))]
    (if (< len cell-width)
      (str (reduce str (repeat (- cell-width len) " ")) e)
      e)))

(defn cell [cell-width product]
  (pad-entry cell-width product))

(defn row [cell-width prime product-row]
  (string/join " "
               (flatten [(cell cell-width prime)
                         vertical-divider
                         (for [product product-row]
                           (cell cell-width product))])))

(defn header [cell-width primes]
  (str
   (repeat-string " " (* 2 cell-width))
   (string/join " "
                (map #(cell cell-width %) primes))))

(defn format-prime-table [{:keys [primes products]}]
  (let [cell-width (->> products
                          flatten
                          (apply max)
                          str
                          count)]
    (loop [accum [(header cell-width primes)
                   (header-divider cell-width (count primes))]
           primes primes
           products products]
      (if (or (empty? primes)
              (empty? products))
        (string/join "\n" accum)
        (recur
         (conj accum (row cell-width (first primes) (first products)))
         (rest primes)
         (rest products))))))

"        2   3   5   7  11  13  17  19  23  29"
"    -----------------------------------------"
"  2 |   4   6  10  14  22  26  34  38  46  58"
"  3 |   6   9  15  21  33  39  51  57  69  87"
"  5 |  10  15  25  35  55  65  85  95 115 145"
"  7 |  14  21  35  49  77  91 119 133 161 203"
" 11 |  22  33  55  77 121 143 187 209 253 319"
" 13 |  26  39  65  91 143 169 221 247 299 377"
" 17 |  34  51  85 119 187 221 289 323 391 493"
" 19 |  38  57  95 133 209 247 323 361 437 551"
" 23 |  46  69 115 161 253 299 391 437 529 667"
" 29 |  58  87 145 203 319 377 493 551 667 841"
